README.md
# 1. Install
After downloading the source code, you will need to do some initial configuration.

You can download the source code here: https://github.com/careerpathways/Career-Pathways-Web-Tool


## Configure Application
* Most site configuration is performed in https://github.com/careerpathways/Career-Pathways-Web-Tool/tree/master/www/include the include folder on the site including:
    *  Database connection settings
    *  Site features
    *  Site name
    *  Site email settings
    *  Language files
    *  Site cache and asset folders
    *  Other environment variables
    *  Recaptcha keys
* Make sure to update email templates by visiting (your domain)/a/emailcontents.php
* You can update counties that the system uses by editing and executing https://github.com/careerpathways/Career-Pathways-Web-Tool/blob/master/scripts/sql/updates/r564_adds_counties_and_fixes_schools.sql after the Database has been constructed (below).


## Configure .htaccess file
Copy the `www/htaccess` file to `www/.htaccess` and modify it to point to your `www/include` and `common` folders.


## Set Up Database
Configure the database username and password in your `www/include/*.settings.php` file. If you can't create your own file, you will have to edit the `default.settings.php` file directly, just make sure not to commit that file back to the repository.

Install the database tables from the `scripts/sql/schema.sql` file provided. After the tables are created, run the `scripts/sql/initdata.sql` file to insert some default data.

Make sure to run updates in scripts/sql to get your database up to date.


## Install Dependencies
From the project root, run `composer install`



# 2. Start
## Development
### Requirements
* [Vagrant](https://www.vagrantup.com/)
* [Composer](https://getcomposer.org/)

Run `vagrant up` from `www/Vagrant/`, then visit http://localhost:8080 in the browser, and you should be presented with a login prompt.

The default user account is **admin@example.com** and a password of **1234**.

## Live Server
### Requirements
The Career Pathways Web Tool is best suited for a Linux web server running Apache, PHP, and MySQL. It can also run on a Windows web server, but has not yet been implemented in Windows on a production system. The following software is required for Linux and Windows installations:

* Apache 2.0 web server with mod_rewrite
* PHP 5.2 or later
* MySQL 5.0 or later
* [Composer](https://getcomposer.org/)
* There are several modules that should be installed. The easiest way to find these modules is by reviewing the Vagrant bootstrap.sh script that is available on github https://github.com/careerpathways/Career-Pathways-Web-Tool/blob/master/www/Vagrant/bootstrap.sh.
* The software requires write permission to several directories. The user running Apache will need write access to the "cache" folder and the "assets" folder.

*Note: It may be possible to run the software under a web server other than Apache, however this has not yet been attempted.*

### Live Server Additional Config
#### Apache
Configure apache to serve files from the "www" folder in the project. This is usually accomplished with a VirtualHost definition like the following.

```
<VirtualHost *:80>
	ServerName test.ctepathways.org

	DocumentRoot /web/test.ctepathways.org/www
	<Directory /web/test.ctepathways.org/www>
                Options +FollowSymLinks
                AllowOverride All
                allow from all
	</Directory>
</Virtualhost>
```

If you have the SetEnv module installed, add this line inside the Directory block:

    SetEnv CONFIG_FILE my.settings.php

Change the filename to whatever you want, and copy www/include/default.settings.php to the filename you set here.


#### Filesystem
If it doesn't already exist, create a `cache` folder next to the `www` and `scripts` folders, and make it writable by apache.
If it doesn't already exist, create a `asset` folder next to the `www` and `scripts` folders, and make it writable by apache.

```
drwxr-xr-x  3 apache   apache    4096 2011-03-24 14:35 cache
drwxr-xr-x  3 apache   apache    4096 2011-03-24 14:35 asset
drwxr-xr-x  7 apache   apache    4096 2011-03-24 14:15 common
drwxr-xr-x  2 apache   apache    4096 2011-03-24 14:15 examples
drwxr-xr-x  3 apache   apache    4096 2011-03-24 14:15 scripts
drwxr-xr-x 12 apache   apache    4096 2011-03-24 14:29 www
```


# Additional Information
## PHP Includes
This project is set up to use php includes and modifies the include path. See `www/Vagrant/bootstrap.sh` where Apache virtual sites are defined for a line similar to `php_value include_path \".:/home/project/$projectName/www/include:/home/project/$projectName/common\"`

Notice that now, when you see `include('x.php');` in PHP, these paths will be checked as well.


# Release Process
* Update `(core)/common/version.php` with appropriate version number.
* *Add config and release notes to `(core)/CHANGELOG.md`*
* Tag the release, e.g. `4.3.0`


CHANGELOG.md
# Career Pathways Web Tool Release Log
*(Since September 17, 2015)*

## 4.5.0 (June 30, 2019)
### Config
* None

### Release Notes
* Upgrade TinyMCE to version 5.


## 4.4.0 (June 25, 2019)
### Config
* None

### Release Notes
* Allow users to resize and move the content editor.


## 4.3.7 (May 25, 2018)
### Config
* None

### Release Notes
* DB update function will now include all values passed to it.


## 4.3.6 (May 21, 2018)
### Config
* None

### Release Notes
* When saving organization, prevent empty fields from causing an SQL error.


## 4.3.2 (August 18, 2017)
### Config
* None

### Release Notes
* Add CHANGELOG.md to project.
* Include Yahoo YUI files within core.
* Update licensing.


## 4.3.1 (June 7, 2017)
### Config
* None

### Release Notes
* Update Readme
* Do not auto-expand embedded drawing container height if height is defined.


## 4.3.0 (May 9, 2017)
### Config
* Execute `(core)/scripts/sql/updates/*_4.3.0.sql`

### Release Notes
* Update PDF and ADA Accessible links in top-right of drawings.
* Properly case words when importing program titles via APN.
* OLMIS Links titles are now acquired from drawing content, no longer depend on Quality Info website.
* Optionally remove roadmap title.
* Break chview.js and chadmin.js into separate files to make debugging easier.
* Fix bug where background color is lost on innerRectangle re-paint.
* Add Approved Program Name report.
* Add totals to APN report.
* Fix vertical alignment of content when hiding roadmap content box title.
* Fix PDF/IE rendering.
* Update verbiage on OLMIS links and add id.
* Fix roadmaps are not displaying at their full height when embedded.
* Make roadmaps canvas twice as tall.
* Add icons, formatting, and small html tweaks for PDF and text only links.
* Adds unit tests.
* Exclusion/exception lists for roadmap drawings now work with the DB.
* The report "Provide a quick breakdown on the types of POST Views" now has subtotals for HS, CC and other org.
* Update "POST report" page for an easier time navigating content.
* Update embed code for roadmap drawings so there is no scrollbar and the iframe takes the height of the drawing.
* Fix composer should not install php unit in non-dev environments.
* PDF exports should not ever contain the links to the pdf and text only versions of drawings.
* Fix PDF link in POST drawings.
* Fix Image library not loading buckets.
* Org admins of other orgs should be able to create post views.
* Anyone with access to a post view should be able to get the embed code.
* Other org users should see the new OLMIS text.
* Help desk email reply-to is now the form-filler's email address when available.
* Remove test embed code for drawings.
* Fixes regex bug affecting alt text entries.
* Adds font choice to Roadmap and POST Drawings.
* Fixes Schools information not displaying correctly on create school page.

## 4.2.x (August 24, 2016)
### Config
* Execute `(core)/scripts/sql/updates/d20160621_add_show_updated_column.sql`
* Execute `(core)/scripts/sql/updates/d20160623_add_alt_column_to_assets_table.sql`
* Execute `(core)/scripts/sql/updates/d20160706_4.2.0.sql`

### Release Notes
* Adds ability to turn "Updated" stamp on/off for published Roadmap and POST Drawings.
* Adds variable connection and line thickness to Roadmap drawings. Light, Medium and Heavy are now available.
* Editors can choose dashed lines as a connection type when editing Roadmap Drawings.
* Site admins can create abbreviations for Degree Types, and Views will utilize them for new tab names.
* Admins can see all of the image info in Image Library.
* Adds cache-invalidation for Image Library files.
* Users can now  enter custom alt text for images.
* Fix Image library was showing the school the asset belonged to. It now shows the school that the asset's creator belongs to.


## 4.1.5 (May 19, 2016)
### Config
* None

### Release Notes
* Updates image library to have a grey background for image assets, so white images with transparency can be seen in all steps.
* Updates wording during image replacement.
* Fixes image replacement in image library across different buckets. Improves Move behavior as well.
* Changed permissions for webmaster to match those of OrgAdmin.

## 4.1.4a (April 19, 2016)
### Config
* None

### Release Notes
* Moves the Image Library “move” feature up above images, so it’s not too far off the screen when used.
* Updates permissions model for the Image Library. “If you uploaded it, you can modify it.”
* Updated help sidebar links styles to be consistent in Roadmap and POST edit screens.

## 4.1.4 (March 24, 2016)
### Config
* t192 - It might be necessary to tweak css on a per-project basis, for this to take effect. I don't think so but be aware.

### Release Notes
* Adds ability to move images to different buckets in the Image Library. [LL WA t72]
* Fixes Approved Program Name not showing up when adding a drawing to a view. [LL OR t193]
* Makes tools sidebar link styles consistent with other sidebar links. [LL OR t195]
* Updates default image library bucket to be user's school, except state admins (who default to the site-wide bucket). [LL WA t74]
* Updates post views, so Community College section tab names have the format: "DEGREE TYPE: POST Drawing Title" by default. High School tab names are unchanged. [LL OR t166]

### Additional Info
* When moving images in the image library:
* A user can move an image out of a bucket if they have the delete permission for that bucket (since they're essentially removing it from that bucket).
* A user will only see buckets that they are allowed to write TO, excluding the current bucket  (obviously). This causes a case where there are no buckets to move the image to, and that's expected/desired as far as I'm aware. I added a message that says something like "Sorry, there are no buckets available to move the image to".



## 4.1.3b (March 16, 2016)
### Config
* None

### Release Notes

common/inc.php
<?php
date_default_timezone_set('America/Los_Angeles');
define('DIR_CORE', dirname(__dir__));

include("version.php");
include("defines.php");

include("database.inc.php");
include("template.inc.php");
include("sitesettings.inc.php");

if(array_key_exists('CONFIG_FILE', $_SERVER))
	include($_SERVER['CONFIG_FILE']);
else
	include("default.settings.php");
	
define('CPUSER_HIGHSCHOOL', 8);
define("CPUSER_STAFF", 16);
define("CPUSER_SCHOOLADMIN", 64);
define("CPUSER_WEBMASTER", 32);
define("CPUSER_STATEADMIN", 127);

include('site-template.inc.php');
include('formatting-functions.inc.php');

include("general.inc.php");
include("admin_inc.php");
include("json_encode.php");
include('Cycler.php');

include('Localize.php');
include('lang/' . $SITE->lang_file . '.php');

/*
 * Is this even used anywhere?
 */
function logmsg($message) {
	global $SITE;
	$fp = fopen($SITE->cache_path('log') . 'log.txt', 'a');
	fwrite($fp, date("Y-m-d H:i:s").' '.$_SERVER['REMOTE_ADDR'].' ['.$_SERVER['PHP_SELF'].'] u:'.$_SESSION['user_id'].' '.$message."\n");
	fclose($fp);
}

function l($str=NULL)
{
	static $loc;
	if(!isset($loc))
		$loc = new Localize_Strings();
	
	if($str === NULL)
		// return the localize object so methods can be run off it directly
		return $loc;
	else
		// default to lookup up a string from the get() method
		return $loc->get($str);
}

function IsLoggedIn() {
	if( $_SESSION['user_level'] > -1 ) {
		return TRUE;
	} else {
		return FALSE;
	}
}

function IsGuestUser() {
	return array_key_exists('email', $_SESSION) && $_SESSION['email'] == 'guest';
}

function RandPass($len = 8){
	$str = "";
    for($i=1;$i<=$len;$i++) $str .= base_convert(rand(0,15),10,16);
    return strtoupper($str);
}


function RequireLogin() {
	if( $_SESSION['user_level'] == -1 ) {
		header("Location: /a/login.php?next=".urlencode($_SERVER['REQUEST_URI']));
	}
}

function CanEditOtherSchools() {
	return $_SESSION['user_level'] == CPUSER_STATEADMIN;
}


function CreateDrawingCodeFromTitle($title, $school_id, $drawing_id=0, $mode='pathways') {
global $DB;
	// replace spaces with underscores
	$dirty_code = preg_replace('/\s+/','-',strtolower($title));

	// remove any character that is not a letter or number
	$clean_code = CleanDrawingCode($dirty_code);

	// remove any duplicate underscores
	$clean_code = preg_replace('/_+/','-',$clean_code);

	// get the coded version of the school abbreviation (removes spaces)
	$school_abbr = CleanDrawingCode($DB->GetValue('school_abbr','schools',$school_id));

	// this is the ideal code for this drawing. but it may already exist in the database
	$proposed_code = strtolower($school_abbr.'-'.$clean_code);

	// look for conflicting codes
	$code = $proposed_code;
	while( DrawingCodeAlreadyExists($code, $drawing_id, $mode) ) {
		// keep trying new codes until we get one that is unique
		$code = $proposed_code.'-'.rand(100,999);
	}

	return $code;
}

function CleanDrawingCode($code) {
	return trim(strtolower(preg_replace('/[^a-z0-9]+/i','-',$code)), ' -');
}

function DrawingCodeAlreadyExists($code, $drawing_id, $mode) {
global $DB;
	$num = $DB->SingleQuery("SELECT COUNT(*) AS num FROM ".($mode=='pathways'?'drawing_main':'post_drawing_main')."
		WHERE code='".$code."'
		AND id != ".$drawing_id);
	return $num['num'] == 1;
}

function GetDrawingInfo($drawing_id, $type='pathways') {
global $DB;
	if( $type == 'pathways' ) {
		$drawing = $DB->SingleQuery("SELECT drawing_main.*, drawings.*, 
				drawings.id AS drawings_id, sk.title AS skillset, school_name,
				IF(drawing_main.name='', p.title, drawing_main.name) AS full_name
			FROM drawing_main
			JOIN drawings ON drawings.parent_id=drawing_main.id
			LEFT JOIN oregon_skillsets AS sk ON sk.id = drawing_main.skillset_id
			LEFT JOIN schools ON drawing_main.school_id=schools.id
			LEFT JOIN programs AS p ON drawing_main.program_id=p.id
			WHERE drawings.id=".$drawing_id);
	} elseif( $type == 'post' ) {
		$drawing = $DB->SingleQuery("SELECT post_drawing_main.*, post_drawings.*, post_drawings.id drawings_id, school_name
			FROM post_drawing_main, post_drawings, schools
			WHERE post_drawings.parent_id=post_drawing_main.id
			AND post_drawing_main.school_id=schools.id
			AND post_drawings.id=".$drawing_id);
	}
	return $drawing;
}

function GetSchoolName($school_id)
{
	global $DB;
	return $DB->GetValue('school_name', 'schools', intval($school_id));
}


function GetAssociatedDrawings($drawing_id, $mode='connections', $type=null)
{
	global $DB;
	$drawing_id = intval($drawing_id);

	if( $mode == 'connections' )
	{
		$type = $DB->GetValue('type', 'post_drawing_main', $drawing_id);
		return $DB->VerticalQuery('SELECT * FROM post_conn WHERE '.($type=='HS'?'hs':'cc').'_id='.$drawing_id, ($type=='HS'?'cc':'hs').'_id');
	}
	else
	{
		return $DB->VerticalQuery('SELECT post_id
									FROM vpost_links AS v
									JOIN post_drawing_main AS d ON v.post_id=d.id
									WHERE type="'.$type.'" AND vid='.$drawing_id, 'post_id');
	}
}

function GetHSAffiliations($cc_id)
{
	global $DB;
	return $DB->MultiQuery('SELECT s.*, a.id AS af_id
		FROM hs_affiliations AS a
		JOIN schools AS s on s.id=hs_id
		WHERE cc_id='.$cc_id.'
		ORDER BY school_name');
}

$getLinkHeaders_cache = array();

function getLinkHeaders($link)
{
    //print "processingg: {$link}\n";
    global $getLinkHeaders_cache;
    if (!isset($getLinkHeaders_cache)) {
      $getLinkHeaders_cache = array();
    }
    if (isset($getLinkHeaders_cache[$link])) {
      //print "using cache\n";
      return $getLinkHeaders_cache[$link];
    }
    
    $returnArray = array(
        'connected' => false,
        'http_code' => 404,
        'url' => $link,
        'redirect_url' => '',
        'error' => 'Empty Link',
    );
    if (empty($link)) {
        //print "got empty!!!\n";
        $getLinkHeaders_cache[$link] = $returnArray;
        return $returnArray;
    }

    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $link);
    curl_setopt($ch, CURLOPT_HEADER, true);
    curl_setopt($ch, CURLOPT_NOBODY, true);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'HEAD');
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
    curl_setopt($ch, CURLOPT_TIMEOUT, 5);
    curl_setopt($ch, CURLOPT_USERAGENT, "Mozilla/4.0 (compatible; MSIE 5.01; Windows NT 5.0)");

    $curlReturn = curl_exec ($ch);
    if(curl_errno($ch)) {
        $returnArray['error'] = curl_error($ch);
        $getLinkHeaders_cache[$link] = $returnArray;
        return $returnArray;
    }
    
    $info = curl_getinfo($ch);
    curl_close ($ch);

    $returnArray =  array(
      'connected' => true,
      'http_code' => $info['http_code'],
      'url' => $info[$link],
      'redirect_url' => $info['redirect_url'],
    );
    $getLinkHeaders_cache[$link] = $returnArray;
    return $returnArray;

}

/**
 * Full base URL of this server.
 * Supports non-standard ports.
 * @todo  move this function to common.
 * @return string - Entire base url with protocol. No trailing slash.
 */
function getBaseUrl(){
	$isSecure = (!empty($_SERVER['HTTPS']) && $_SERVER['HTTPS'] !== 'off') || $_SERVER['SERVER_PORT'] == 443 || $_SERVER['SERVER_PORT'] == 8443;
	$protocol = '//';
	$_baseUrl = $protocol . $_SERVER['SERVER_NAME'];
	if($_SERVER['SERVER_PORT'] != 80 && $_SERVER['SERVER_PORT'] != 443){
		$_baseUrl .= ':' . $_SERVER['SERVER_PORT'];
	}
	return $_baseUrl;
}

/**
 * Tries to find the external link to the specified drawing in the external_links table.
 * Filters out things like file:/// links and known testing URLs.
 * 
 * @param int $drawing_id
 * @param string $type 'post' or 'pathways'
 * @return string
 */
function getExternalDrawingLink($drawing_id, $type)
{
	global $DB;
	$time_start = microtime(true);
	
	$drawing_id = intval($drawing_id);
	if($type !== 'pathways' && $type !== 'post_views'){
		$type = 'post';
	}
	$link = $DB->SingleQuery('
		SELECT url
		FROM external_links 
		WHERE drawing_id = ' . $drawing_id . ' 
			AND type = "' . $type . '"
			AND `primary` = 1
		');

	// If there was no "best link" marked in the database, find one and mark it
	$url = FALSE;
	if ($link == FALSE || count($link) == 0) {
		$links = getExternalDrawingLinks($drawing_id, $type);
		if(array_key_exists(0, $links))
		{
			$DB->Update('external_links', array('`primary`'=>1), $links[0]['id']);
			$url = $links[0]['url'];
		}
	} else {
		$url = $link['url'];
	}
	if ($url!==FALSE) {
		//test the link
		$result = getLinkHeaders($url);
		if ($result['connected']) {
		    if($result['http_code'] == 200){
			    //SUCCESS: Do nothing, the url has been checked and will be returned.
			    //print($result[0]."     ".$url.PHP_EOL);
		    } elseif($result['http_code'] == 404 || $result['http_code'] == 403) {
			    //FAIL: Not found
			    $url = FALSE;
		    } else {
			    //FAIL: Some header other than the normal 200 status was returned.
			    $url = FALSE;
		    }
		} else {
			//FAIL: We could not even connect to the URL
			$url = FALSE;
		}
		
		
		/*
		$result = @get_headers($url);
		if(!empty($result) && stripos($result[0],"200 OK")!==false){
			//SUCCESS: Do nothing, the url has been checked and will be returned.
			//print($result[0]."     ".$url.PHP_EOL);
		} else {
			//FAIL: Some header other than the normal 200 status was returned.
			$url = FALSE;
		}
		*/
	}
	return $url;
}

function getExternalDrawingLinks($drawing_id, $type, $orderBy='counter DESC')
{
	global $DB;
	
	$drawing_id = intval($drawing_id);
	$type = ($type == 'pathways' ? 'pathways' : 'post');
	
	$links = $DB->MultiQuery('
		SELECT id, url, counter, last_seen, `primary`
		FROM external_links 
		WHERE drawing_id = ' . $drawing_id . ' 
			AND type = "' . $type . '"
			AND counter > 1
		ORDER BY ' . $orderBy, 'url');
	$exclude = $DB->VerticalQuery('SELECT pattern FROM external_link_exclude', 'pattern');
	
	$bestLinks = array();
	// start with the least likely, refining the search as we look through the list
	foreach($links as $link)
	{
		// if the link matches any of the excluded patterns, skip it
		foreach($exclude as $e)
			if(preg_match($e, $link['url']))
				continue 2;

		$bestLinks[] = $link;
	}
	return $bestLinks;
}


function drawing_not_found($type, $drawing_id=0, $version_id=0, $drawing_code='', $version=0)
{
	global $DB;
	
	$data = array(
		'date' => date('Y-m-d H:i:s'),
		'type' => $type,
		'request_uri' => $_SERVER['REQUEST_URI'],
		'query_string' => $_SERVER['QUERY_STRING'],
		'external_url' => array_key_exists('HTTP_REFERER', $_SERVER) ? $_SERVER['HTTP_REFERER'] : '',
		'drawing_id' => $drawing_id,
		'version_id' => $version_id,
		'version' => $version,
		'drawing_code' => $drawing_code,
		'counter' => 1
	);

	// Check if we already have a record of this request URI
	$check = $DB->SingleQuery('SELECT COUNT(1) AS num FROM external_link_errors WHERE request_uri = "' . $data['request_uri'] . '"');
	if($check['num'] > 0)
	{
		$DB->Query('UPDATE external_link_errors 
			SET counter = counter + 1, date = NOW()
			WHERE request_uri = "' . $data['request_uri'] . '"');
	}
	else
	{
		$DB->Insert('external_link_errors', $data);
	}
	
	header('HTTP/1.0 404 Not Found');
	echo 'Not found';
	die();
}


function drawing_sort_by_version($a,$b) {
	return $a['version_num'] < $b['version_num'];
}


function getbrightness($hex) {
	$hex = strtolower($hex);
	$r = hexdec(substr($hex, 0, 2));
	$g = hexdec(substr($hex, 2, 2));
	$b = hexdec(substr($hex, 4, 2));
	$brightness = $r + $g + $b;
	return $brightness;
}

function getdominantcolor($hex) {
	$hex = strtolower($hex);
	$r = hexdec(substr($hex, 0, 2));
	$g = hexdec(substr($hex, 2, 2));
	$b = hexdec(substr($hex, 4, 2));

	$max = max($r, $g, $b);
	if( ($r == $g) && ($g == $b) ) {
		return 'grey';
	}
	if( $r == $max ) {
		return 'r';
	}
	if( $g == $max ) {
		return 'g';
	}
	if( $b == $max ) {
		return 'b';
	}
}

/**
 * If $value is not null, returns $value. Otherwise returns $default
 */
function dv($value, $default='')
{
	return ( $value ? $value : $default );	
}

/*
 * Return a handle to the $DB object, easier than using `global $DB` everywhere
 */
function db()
{
	return $GLOBALS['DB'];
}

function varDumpString($var)
{
	ob_start();
	var_dump( $var );
	$out = ob_get_clean();
	return $out;
}